### PR TITLE
Add extract session only endpoint

### PR DESCRIPTION
**Pull request by Bibob7**
## Description

I added a new endpoint /session that will always response with 202. Depending on the oauth2-proxy config, the response will contain specified headers for authentication.

## Motivation and Context

We want to use oauth2-proxy only to extract the access/id_token from available session data.
If no session exists, it just also forward the request to the upstream service, because we use Envoy to implement further authentication and authorization policies. In some cases authentication is optional, in this case the application decides what to respond. In this way we only pass valid or no token to the upstream service.

These changes are a first naive approach to solve the problem from [#872](https://github.com/oauth2-proxy/oauth2-proxy/issues/872).

## How Has This Been Tested?

I simply called the new session endpoint with and without an existing session.

## Checklist:

- [X] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: oauthproxy.go [hunk 1]
```diff
@@ -887,6 +891,18 @@ func (p *OAuthProxy) AuthenticateOnly(rw http.ResponseWriter, req *http.Request)
 	rw.WriteHeader(http.StatusAccepted)
 }
 
+func (p *OAuthProxy) ExtractSessionOnly(rw http.ResponseWriter, req *http.Request) {
```
Comment (Bibob7) [inline oauthproxy.go:894]: Instead we could also extend the AuthenticateOnly method to handle errors in two ways. What would you prefer?

**Comment by NickMeves** [comment]
What's the need for this instead of using the existing `/oauth2/auth` endpoint (which returns 202) and it will return any session details you need in the `X-Auth-Request-*` headers if you enable them in the configuration options?

I would aim to tweak the existing `/oauth2/auth` endpoint if it needs to be extended in a minor way for a common use case.

**Comment by Bibob7** [comment]
> What's the need for this instead of using the existing `/oauth2/auth` endpoint (which returns 202) and it will return any session details you need in the `X-Auth-Request-*` headers if you enable them in the configuration options?

There is no need for this. It was just my first idea to clearly distinguish these two behaviors. I am open for alternatives.
 
> I would aim to tweak the existing `/oauth2/auth` endpoint if it needs to be extended in a minor way for a common use case.

 This is totally fine for me.

> What's the rationale to forward to the upstream in the failure case?

Our idea is to do the validation of the tokens via a sidecar directly at the service and to define different policies. But it should also be possible that authentication is optional for some endpoints. So a token is not always mandatory. For frontend -> backend communication, however, this requires that the token from the session cookie is first decrypted with the oauth2-proxy Service and then extracted for the header.

> It would get a `/oauth2/(?:session|auth)` path which is unusual -- we've traditionally reserved everything under the path 
> prefix (default `/oauth2`) to not proxy to the upstream.

Does that mean we would have to make the basic forwarding behavior configurable?
With a certain configuration, I could imagine that if ErrNeedsLogin occurs, we will forward directly to the upstream.

**Comment by NickMeves** [comment]
Do you by chance have a good reference document for how Envoy/Istio builds an architecture with `OAuth2-Proxy` so I can visualize this better (I haven't used them myself).

I want to make sure I understand if it is a `subrequest` type auth architecture like  Kubernetes ingress annotations checking an `auth` endpoint for `2XX` or `4XX` and then the ingress proxying the request to the backend. Is it this, it this the reason `Envoy` users have to do the `static://200` for the `upstream` configuration?

Or if it is an inline proxy, where the communication path is `client` -> (other proxies in chain) -> `OAuth2-Proxy` -> `Backend App`.

**Comment by Bibob7** [comment]
> Do you by chance have a good reference document for how Envoy/Istio builds an architecture with `OAuth2-Proxy` so I can visualize this better (I haven't used them myself).
> 
> I want to make sure I understand if it is a `subrequest` type auth architecture like Kubernetes ingress annotations checking an `auth` endpoint for `2XX` or `4XX` and then the ingress proxying the request to the backend. Is it this; is this the reason `Envoy` users have to do the `static://200` for the `upstream` configuration?
> 
> Or if it is an inline proxy, where the communication path is `client` -> (other proxies in chain) -> `OAuth2-Proxy` -> `Backend App`.
> 
> In my head, those are 2 distinct architectures. But what you've described is somehow merging the two paradigms and I'm having trouble visualizing.

In my use case I want to use Nginx with external authentication. About Envoy I'm not sure, but I would expect that it works in the same way as Nginx Ingress ext. auth. At least it sounds like this: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter

**Comment by NickMeves** [comment]
So for the nginx with external authentication, the OAuth2-Proxy is essentially adjacent to the proxied request route chain (as a subrequest from the nginx layer). And in these scenarios, an `upstream` doesn't matter since it mostly exists for the `/oauth2/auth` endpoint to return 2XX or 4XX

So how would it sometime forwarding to an upstream play into this architecture? Do you have other things hooking into it on the side?

I'm still struggling to visualize 😅  -- If you have some rudimentary ascii art diagrams that might help.

But this is sounding like a corner case architecture that we don't want to include in the main branch at risk of overly complicating the codebase and confusing new end users (nginx auth subrequest is confusing enough and that whole section of the docs needs an overhaul).

**Comment by NickMeves** [comment]
Oh - maybe I'm confused by this line in the description:

> If no session exists, it just also forward the request to the upstream service, because we use Envoy to implement further authentication and authorization policies

By `it` you actually don't mean `OAuth2-Proxy` but your proxy (nginx, envoy, whatever)?

**Comment by Bibob7** [comment]
> Oh - maybe I'm confused by this line in the description:
> 
> > If no session exists, it just also forward the request to the upstream service, because we use Envoy to implement further authentication and authorization policies
>
> By `it` you actually don't mean `OAuth2-Proxy` but your proxy (nginx, envoy, whatever)?

Sorry for the confusion. The forwarding is done by the nginx ingress. Envoy we use as sidecar to implement authorization policies. For example checking, if a specific claim in the token is set.

**Comment by Bibob7** [comment]
> So for the nginx with external authentication, the OAuth2-Proxy is essentially adjacent to the proxied request route chain (as a subrequest from the nginx layer). And in these scenarios, an `upstream` doesn't matter since it mostly exists for the `/oauth2/auth` endpoint to return 2XX or 4XX
> 
> So how would it sometime forwarding to an upstream play into this architecture? Do you have other things hooking into it on the side?
> 
> I'm still struggling to visualize 😅 -- If you have some rudimentary ascii art diagrams that might help.
> 
> But this is sounding like a corner case architecture that we don't want to include in the main branch at risk of overly complicating the codebase and confusing new end users (nginx auth subrequest is confusing enough and that whole section of the docs needs an overhaul).

In the end we want to use oauth2-proxy just to extract the access or id token from the cookie and refresh it if necessary. It will not act as a proxy.

**Comment by NickMeves** [comment]
So this feels like a corner case architecture that advertising too much would more lead to misconfigurations and security bugs for standard users.

BUT -- were we to implement it, I think I understand what you need.

Does this work better as optional querystring parameters on the `/oauth2/auth` endpoint where you can customize the success/failure error codes? And existing header functionality would persist (success gets session headers, failure gets no headers added).

I'm looking at utilizing a similar concept in this PR for group authorization: https://github.com/oauth2-proxy/oauth2-proxy/pull/849

**Comment by Bibob7** [comment]
> Does this work better as optional querystring parameters on the `/oauth2/auth` endpoint where you can customize the success/failure error codes? And existing header functionality would persist (success gets session headers, failure gets no headers added).

I like the idea!

**Comment by JoelSpeed** [comment]
If I've understood correctly, and you are using nginx auth request, can you not just ignore the reponse code of the auth request to achieve your desired outcome?

Normally you have something like:

```
auth_request /oauth2/auth
error_page 401 /oauth2/sign_in
```

If you were to just remove the `error_page`, part, nginx shouldn't do anything and can carry on to proxy to your upstream right?

**Comment by JoelSpeed** [comment]
For reference, I got something working that I believe is what you want using the following nginx snippet
```
  location / {
    auth_request /internal-auth/oauth2/auth;
    error_page 401 =200 @upstream;

    proxy_pass http://httpbin;
  }

  location @upstream {
    proxy_pass http://httpbin;
  }
```

Alternatively

```
  location / {
    auth_request /internal-auth/oauth2/auth;
    error_page 401 =200 /internal-upstream/$request_uri;

    proxy_pass http://httpbin;
  }

  location /internal-upstream/ {
    internal;

    proxy_pass http://httpbin/;
  }
```

Basically you override the error page to be a 200 response and then proxy to the upstream as normal, but you have to use a second block for this. If you need additional configuration I suggest using an include statement to ensure the config is common between the two location blocks

**Pull request closed without merging**